juju/juju
# feat: Add support for TopologySpreadConstraints

This change fixes [LP#2039316](https://bugs.launchpad.net/juju/+bug/2039316).

Currently, the only mechanisms available to set topology in k8s are the tag constraints `anti-pod|node|pod`. That enables to have certain configurations, such as enforce that all replicas of a given database land in different nodes or availability zones.

However, during the lifecycle of our application, these pods may migrate between nodes and end up in a configuration that does not respect anymore the original placement constraints, e.g. all database replicas landing in the same AZ now. 

This PR extends existing k8s constraints and adds support for [`TopologySpreadConstraints`](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/). It brings two benefits: (1) the [`TopologySpreadConstraints`](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) is k8s official response to the problem of controlling topology and placement over the lifecycle of applications; and (2) we do not need to manually tweak the statefulset that Juju manages, post original deployment.

The current proposal adds the following tags to CAAS constraints:
* maxSkew: we accept 1 or more pods to co-exist and not respecting the topology constraint at a time
* minDomains: how many domains do we expect the deployment to have (e.g. number of AZs): it allows an extra validation that the environment has the needed domain count
* nodeTaintsPolicy: allows the user to follow with taints besides the topology spread rules
* matchLabel: custom-defined labels, following the same pattern as the anti-pod|pod|node tags

This PR also enforces `nodeAffinityPolicy` to `Honor` and forbids further scheduling if the topology constraints are not satisfied (i.e. `WhenUnsatisfiable` == `DoNotSchedule`).

### How to Use

One example of command:

```
juju deploy postgresql-k8s --constraints="tags=topology-spread.topology-key=kubernetes.io/hostname" --channel=14/edge -n3
```

Will set the `TopologySpreadConstraints` to spread across nodes following the `hostname` tag.

## Checklist



- [x ] Code style: imports ordered, good names, simple structure, etc
- [x ] Comments saying why design decisions were made -> following the same patterns as the original `constraints.go`
- [x ] Go unit tests, with comments saying what you're testing
- [? ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing - could not find any tests for constraints specifically

## QA steps

1. deploy microk8s and enable `dns` and `storage`
2. bootstrap the controller on k8s or add the microk8s as a cloud to the controller
3. Run the steps to deploy:

```
juju add-model postgres
juju create-storage-pool hostpath kubernetes storage-class=microk8s-hostpath
juju deploy postgresql-k8s --constraints="tags=topology-spread.topology-key=kubernetes.io/hostname" --channel=14/edge -n3
```

Once the setup is done, the skewing should kick in (as the current setup is missing `minDomains=3`) and a single unit is successfully created. The setup will look like:
```
$ kubectl get po  -n postgres       
NAME                             READY   STATUS    RESTARTS   AGE    
modeloperator-5c56fdf9fc-vxn47   1/1     Running   0          26m    
postgresql-k8s-0                 2/2     Running   0          26m    
postgresql-k8s-1                 0/2     Pending   0          26m    
postgresql-k8s-2                 0/2     Pending   0          26m    
```

Remaining units should fail with:
```
  Type     Reason            Age    From               Message                                                                                            
  ----     ------            ----   ----               -------                                                                                            
  Warning  FailedScheduling  2m48s  default-scheduler  0/1 nodes are available: 1 node(s) didn't match pod topology spread constraints. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.                                                                                        
  Warning  FailedScheduling  2m45s  default-scheduler  0/1 nodes are available: 1 node(s) didn't match pod topology spread constraints. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.                                                                                        
```

## Documentation changes

TODO: document the new constraints options


## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2039316
